File: NOU/refusInfo/normalize_json/normalize_json.js
const fs = require('fs');
const path = require('path');

// File paths
const inputJsonPath = path.join(__dirname, 'refusInfo.json');
const outputJsonPath = path.join(__dirname, 'refusInfo_normalized.json');

try {
  // Read the JSON file
  const jsonContent = fs.readFileSync(inputJsonPath, 'utf8');
  const data = JSON.parse(jsonContent);
  
  // Normalize each node
  const normalizedNodes = data.nodes.map(node => {
    const normalizedNode = {};
    
    // Keep id, nom, derniere_modif as they are (lien will be moved to links array)
    normalizedNode.id = node.id;
    normalizedNode.nom = node.nom;
    // Extract only the date part from derniere_modif
    if (node.derniere_modif) {
      normalizedNode.derniere_modif = node.derniere_modif.split(' ')[0];
    }
    
    // Handle coordinates - separate altitude and keep lat/long
    if (node.coord) {
      normalizedNode.coord = {
        long: parseFloat(node.coord.long),
        lat: parseFloat(node.coord.lat)
      };
      
      // Add altitude as separate field
      if (node.coord.alt) {
        const altValue = parseInt(node.coord.alt, 10);
        normalizedNode.altitude = altValue === 0 ? null : altValue;
      } else {
        normalizedNode.altitude = null;
      }
    }
    
    // Simplify places - extract valeur
    if (node.places && node.places.valeur !== undefined) {
      normalizedNode.places = node.places.valeur;
    }
    
    // Simplify etat - extract valeur
    if (node.etat && node.etat.valeur !== undefined) {
      normalizedNode.etat = node.etat.valeur;
    }
    
    // Simplify remarque - extract valeur
    if (node.remarque && node.remarque.valeur !== undefined) {
      normalizedNode.remarque = node.remarque.valeur;
    }
    
    // Handle info_comp - simplify nested structures
    if (node.info_comp) {
      normalizedNode.info_comp = {};
      
      // List of fields to simplify (extract valeur)
      const fieldsToSimplify = [
        'manque_un_mur',
        'cheminee', 
        'poele',
        'couvertures',
        'latrines',
        'bois',
        'eau'
      ];
      
      fieldsToSimplify.forEach(field => {
        if (node.info_comp[field] && node.info_comp[field].valeur !== undefined) {
          let value = node.info_comp[field].valeur;
          // Convert numeric strings to numbers
          if (typeof value === 'string' && /^\d+$/.test(value.trim())) {
            value = parseInt(value.trim(), 10);
          }
          normalizedNode.info_comp[field] = value;
        }
      });
      
      // Handle places_matelas - simplify to just the number from nb
      if (node.info_comp.places_matelas) {
        let nbValue = 0; // Default value if no nb found
        if (node.info_comp.places_matelas.nb !== undefined) {
          // Convert to number if it's a numeric string and not empty
          if (typeof node.info_comp.places_matelas.nb === 'string' && 
              node.info_comp.places_matelas.nb.trim() !== '' && 
              /^\d+$/.test(node.info_comp.places_matelas.nb.trim())) {
            nbValue = parseInt(node.info_comp.places_matelas.nb.trim(), 10);
          } else if (typeof node.info_comp.places_matelas.nb === 'number') {
            nbValue = node.info_comp.places_matelas.nb;
          }
        }
        normalizedNode.info_comp.places_matelas = nbValue;
      }
    }
    
    // Keep description as is
    if (node.description !== undefined) {
      normalizedNode.description = node.description;
    }
    
    // Create links array from lien and site_officiel (without keeping original fields)
    normalizedNode.links = [];
    
    // Add lien if exists and is valid
    if (node.lien && typeof node.lien === 'string' && node.lien.trim() !== '') {
      normalizedNode.links.push(node.lien.trim());
    }
    
    // Add site_officiel from info_comp if exists and is valid
    if (node.info_comp && node.info_comp.site_officiel && node.info_comp.site_officiel.url) {
      const siteUrl = node.info_comp.site_officiel.url;
      if (typeof siteUrl === 'string' && siteUrl.trim() !== '') {
        normalizedNode.links.push(siteUrl.trim());
      }
    }
    
    return normalizedNode;
  });
  
  // Create the normalized result
  const normalizedResult = {
    nodes: normalizedNodes
  };
  
  // Write the normalized JSON to file
  fs.writeFileSync(outputJsonPath, JSON.stringify(normalizedResult, null, 2), 'utf8');
  console.log(`Normalization successful! Normalized JSON saved as ${outputJsonPath}`);
  console.log(`Original nodes: ${data.nodes.length}`);
  console.log(`Normalized nodes: ${normalizedNodes.length}`);
  
  // Show a sample of the first normalized node
  console.log('\nSample of first normalized node:');
  console.log(JSON.stringify(normalizedNodes[0], null, 2));
  
} catch (err) {
  console.error(`Error: ${err.message}`);
}

File: ANTIC/utils/normalizeJSON/normalize_json.py
import json
import os

# Load the JSON file
file_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'refusInfoCompleta.json')
with open(file_path, 'r', encoding='utf-8') as f:
    data = json.load(f)

# Function to normalize the structure of a node
def normalize_node(node):
    normalized_node = {
        "id": node.get("id", ""),
        "nom": node.get("nom", ""),
        "lien": node.get("lien", ""),
        "coord": {
            "long": "",
            "lat": ""
        },
        "alt": "",
        "places": "",
        "etat": {
            "valeur": "",
            "id": "ouverture"
        },
        "remarque": "",
        "info_comp": {
            "site_officiel": {
                "nom": "Site Internet",
                "valeur": "0"
            },
            "manque_un_mur": "0",
            "cheminee": "0",
            "poele": "0",
            "couvertures": "0",
            "places_matelas": "0",
            "latrines": "0",
            "bois": "0",
            "eau": "0"
        },
        "description": ""
    }
    
    # Handle coordinates
    if "coord" in node:
        if "long" in node["coord"]:
            normalized_node["coord"]["long"] = node["coord"]["long"]
        if "lat" in node["coord"]:
            normalized_node["coord"]["lat"] = node["coord"]["lat"]
    
    # Handle altitude
    if "alt" in node:
        normalized_node["alt"] = node["alt"]
    elif "coord" in node and "alt" in node["coord"]:
        normalized_node["alt"] = node["coord"]["alt"]
    
    # Handle places
    if "places" in node:
        if isinstance(node["places"], dict) and "valeur" in node["places"]:
            normalized_node["places"] = node["places"]["valeur"]
        else:
            normalized_node["places"] = str(node["places"])
    
    # Handle etat (state)
    if "etat" in node:
        if isinstance(node["etat"], dict):
            if "valeur" in node["etat"]:
                normalized_node["etat"]["valeur"] = node["etat"]["valeur"]
            if "id" in node["etat"]:
                normalized_node["etat"]["id"] = node["etat"]["id"]
    
    # Handle remarque (remark)
    if "remarque" in node:
        if isinstance(node["remarque"], dict) and "valeur" in node["remarque"]:
            normalized_node["remarque"] = node["remarque"]["valeur"]
        else:
            normalized_node["remarque"] = str(node["remarque"])
    
    # Handle info_comp (complementary information)
    if "info_comp" in node:
        # Site officiel
        if "site_officiel" in node["info_comp"]:
            normalized_node["info_comp"]["site_officiel"] = node["info_comp"]["site_officiel"]
        
        # Process boolean or object fields and convert to simple strings
        for field in ["manque_un_mur", "cheminee", "poele", "couvertures", "latrines", "bois", "eau"]:
            if field in node["info_comp"]:
                if isinstance(node["info_comp"][field], dict) and "valeur" in node["info_comp"][field]:
                    normalized_node["info_comp"][field] = node["info_comp"][field]["valeur"]
                else:
                    normalized_node["info_comp"][field] = str(node["info_comp"][field])
        
        # Handle places_matelas specifically
        if "places_matelas" in node["info_comp"]:
            if isinstance(node["info_comp"]["places_matelas"], dict) and "valeur" in node["info_comp"]["places_matelas"]:
                normalized_node["info_comp"]["places_matelas"] = node["info_comp"]["places_matelas"]["valeur"]
            else:
                normalized_node["info_comp"]["places_matelas"] = str(node["info_comp"]["places_matelas"])
    
    # Handle description
    if "description" in node:
        normalized_node["description"] = node["description"]
    
    return normalized_node

# Normalize all nodes
normalized_nodes = [normalize_node(node) for node in data["nodes"]]

# Replace the nodes in the original data
data["nodes"] = normalized_nodes

# Write the normalized data back to file
output_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'refusInfoCompleta_normalized.json')
with open(output_path, 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

print(f"Normalized JSON written to {output_path}")
